### PR TITLE
Scc 2910

### PR DESCRIPTION
**Pull request by danamansana**
Currently working on tests but I have some questions:

1. For test 
```
Bib Marc Mapping
       Parse
         should pass over parallel fields with malformed 880 $6 (reverse link
```
it is currently returning `[ null, '' ]`. This fails the test, but is it actually OK?

2. Similarly, for test 
```
 Bib Marc Mapping
       Parallel field extraction
         should skip over malformed 880s:
```

it is currently returning 
```
Statement {
  subject_id: 'b21581489',
  predicate: 'nypl:parallelTitle',
  source_id: 10004,
  source_record_id: '21581489',
  creator_id: 1,
  index: 0,
  object_literal: ' ',
  source_record_path: '880[$u^=245] $a $b' }
```
again this fails, but is it actually OK?

3. Finally, the new test I have written makes sense to me and passes, but `bib.literal('nypl:parallelPlaceOfPublication')` returns `长沙市 :`, which I don't understand.

### CODE REVIEW

File: test/bib-marc-test.js [hunk 1]
```diff
@@ -68,6 +68,7 @@ describe('Bib Marc Mapping', function () {
       // that 1) the malformed back-link is treated as such and 2) its presence
       // does not disrupt other parallel field extraction.
       const parallelPlaceOfPublication = bib.parallel('264', ['a', 'b', 'c', 'f', 'g', 'h', 'k', 'n', 'p', 's'])
+      console.log('parallelPlaceOfPublication: ', parallelPlaceOfPublication)
       assert.strictEqual(parallelPlaceOfPublication.length, 0)
```
Comment (nonword) [inline test/bib-marc-test.js:72]: This evaluates to `[null, '']`. It should probably evaluate to `['', '']` for consistency, but the reason it evaluates to `[null, '']` in this edge case is:
 - SierraRecord.parallel loops over the primary values
 - For each primary value, it looks up the parallel value:
   - If there is no $6 in the primary field, [it returns ''](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L97) (This is the case for the second 264 with $c = "©2015".) 
   - If there is a $6 in the primary but it's malformed, [it returns ''](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L102)
   - It then [queries the record for the parallel value](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L105)
   - The [result of that](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L137) is an array of matching varfield values (which should have a single value)
   - If the highly unusual scenario occurs where the valid-looking parallel link in the primary field doesn't actually refer to any parallel field, the function [happens to return `null`](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L140) This is the case with the second 264, whose $6 is "880-03", implying there should be a 264 with $6 = "264-03", but there is not.

Previously, the policy was to collect parallel values and then remove any blank (falsey, technically) values from the end of that list. Now that you've removed the code that does that, falsey values are left untouched. This is all working correctly, but we should adjust the test to expect the last two values to be falsey/blank.

But I think we should also adjust how we communicate this kind of broken 880 link and return `''` [instead of `null`](https://github.com/NYPL-discovery/discovery-store-poster/blob/c97730b92d82948e9b3638bedc8433cea8ab1a57/lib/models/sierra-record.js#L140) for consistency. I don't see a reason to treat them differently.

File: lib/serializers/bib.js [hunk 2]
```diff
@@ -440,17 +440,21 @@ var fromMarcJson = (object, datasource) => {
...
+    parallelMappings.forEach((fieldMapping) => {
+      fieldMapping.paths.forEach((path) => {
+        const subfields = path.subfields || [];
+        const parallelValues = object.parallel(path.isParallelFor, subfields)
+        if (parallelValues && parallelValues.length > 0) {
+          parallelValues.forEach((parallelValue) => {
+            const sourceRecordPath = `880[$u^=${path.isParallelFor}] ${subfields.map((subfield) => `$${subfield}`).join(' ')}`
+            builder.add(fieldMapping.pred, { literal: parallelValue || ' ' }, null, { path: sourceRecordPath })
+          })
+        }
```
Comment (nonword) [inline lib/serializers/bib.js:457]: Currently this is looping over each of the "paths", calling `builder.add` as values are found. This works fine when there are parallel values. When a given mapping has no parallel values we end up calling `builder.add` on a bunch of empty values, resulting in a bunch of useless statements (or empty statements at the end of the list). I wonder if we should apply the rtrim here (instead of in the SierraRecord.parallel call). We'd have to change the structure of the above slightly to _collect_ the statements to be added, then rtrim them (remove blank statements from the end), and _then_ call `builder.add` on them (if any remain).. Essentially we have to do that cleanup work either here or in the indexer later..

File: test/bib-marc-test.js [hunk 3]
```diff
@@ -1036,6 +1065,7 @@ describe('Bib Marc Mapping', function () {
           // We have one 245 with $6=880-01
           // The first 880 has a valid reverse link in $6
           // but does not define any other subfields, so no valid value:
+          console.log('nyplParallelTitle: ', bib.statement('nypl:parallelTitle'))
           assert(!bib.statement('nypl:parallelTitle'))
```
Comment (nonword) [inline test/bib-marc-test.js:1069]: This test asserts that there are no `nypl:parallelTitle` statements, when in fact there is one because you've changed the policy (no longer removing trailing empty values). So this test as is will fail unless we start trimming blank statements again (I propose a different, better place to do that above..)

File: package.json [hunk 4]
```diff
@@ -27,7 +27,7 @@
     "standard": "^14.3.4"
   },
   "scripts": {
-    "test": "./node_modules/.bin/mocha",
+    "test": "./node_modules/.bin/standard --env mocha && ./node_modules/.bin/mocha",
```
Comment (nonword) [inline package.json:30]: 👍

**Review by nonword — APPROVED** [review]
Glad this ended up not being _too_ great a hassle. Looks good.
